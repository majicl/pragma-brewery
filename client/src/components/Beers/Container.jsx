import React from 'react';
import { connect } from 'react-redux';
import './Container.scss';

const BeerCard = ({
  name,
  temperature,
  currentTemperature,
  status = Status.NORMAL,
}) => {
  const statusToClass = () => {
    if (status === Status.COLD) {
      return 'cold';
    }
    if (status === Status.HOT) {
      return 'hot';
    }
    return '';
  };

  return (
    <div data-testid="card" className={`card ${statusToClass()}`}>
      <div className="card-body text-center">
        <h5 className="mb-2 text-dark font-weight-normal">{name}</h5>
        <h2 className="mb-4 text-dark font-weight-bold">
          {currentTemperature}
          °C
        </h2>
        <div className="dashboard-progress dashboard-progress-1 d-flex align-items-center justify-content-center item-parent">
          <i className="mdi mdi-oil-temperature icon-md absolute-center text-dark" />
        </div>
        <p className="mt-4 mb-0 status-label" data-testid="status-label">
          {status.description}
        </p>
        <h4 className="mb-0 mt-2 text-dark">
          <span>Condition:</span>
          {temperature.min}
          °C
          <span> - </span>
          {temperature.max}
          °C
        </h4>
      </div>
    </div>
  );
};

const Status = Object.freeze({
  HOT: Symbol('Hot'),
  COLD: Symbol('Cold'),
  NORMAL: Symbol('Normal'),
});

const isOutsideTemperature = (currentTemperature, temperatureRange) =>
  currentTemperature < temperatureRange.min ||
  currentTemperature > temperatureRange.max;

const getTemperatureStatus = (currentTemperature, temperatureRange) => {
  if (!isOutsideTemperature(currentTemperature, temperatureRange)) {
    return Status.NORMAL;
  }
  if (currentTemperature < temperatureRange.min) {
    return Status.COLD;
  }
  return Status.HOT;
};

function mapStatetoProps(state, props) {
  const { outsideOfTemperature } = state.beers;
  const currentTemperature =
    outsideOfTemperature[props.id] || props.currentTemperature;

  return {
    status: getTemperatureStatus(currentTemperature, props.temperature),
    currentTemperature,
  };
}
export default connect(mapStatetoProps)(BeerCard);
