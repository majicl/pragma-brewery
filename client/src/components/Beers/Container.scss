.card {
    background-color: #14eda7 !important;
    box-shadow: 0 0 5px 2px black;
    transition: all 2s;
    .status-label {
        color: black;
        font-weight: bold;
    }
}
.hot {
    background-color: crimson !important;
    box-shadow: 0 0 13px 2px black;
    transition: all 2s;
    .status-label {
        color: white;
        font-weight: bold;
    }
}
.cold {
    background-color: #147fed !important;
    box-shadow: 0 0 13px 2px black;
    transition: all 2s;
    .status-label {
        color: white;
        font-weight: bold;
    }
}
