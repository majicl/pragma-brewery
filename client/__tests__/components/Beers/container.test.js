import React from 'react';
import '@testing-library/jest-dom';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import { configure } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { render } from '@testing-library/react';
import Container from '~/Beers/Container.jsx';

configure({ adapter: new Adapter() });

const defaultProps = {
  id: 1,
  temperature: {
    min: -1,
    max: 3
  },
  currentTemperature: 1
};
const renderWithRedux = (
  children,
  initialState = {
    beers: {
      outsideOfTemperature: {}
    }
  },
  mockedReducer = (state) => state
) => {
  const store = createStore(mockedReducer, initialState);
  const component = <Provider store={store}>{children}</Provider>;
  return {
    ...render(component),
    component
  };
};

describe('<Container />', () => {
  it('matches snapshot on responsive', () => {
    const { container } = renderWithRedux(<Container {...defaultProps} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Status Sould be Normal', () => {
    const { getByTestId } = renderWithRedux(<Container {...defaultProps} />);
    expect(getByTestId('status-label')).toHaveTextContent('Normal');
  });

  it('Status Sould be Normal: -1', () => {
    const { getByTestId } = renderWithRedux(<Container {...defaultProps} currentTemperature={-1} />);
    expect(getByTestId('status-label')).toHaveTextContent('Normal');
  });

  it('Status Sould be Normal: 3', () => {
    const { getByTestId } = renderWithRedux(<Container {...defaultProps} currentTemperature={3} />);
    expect(getByTestId('status-label')).toHaveTextContent('Normal');
  });

  it('Status Sould be Hot', () => {
    const { getByTestId } = renderWithRedux(<Container {...defaultProps} currentTemperature={4} />);
    expect(getByTestId('status-label')).toHaveTextContent('Hot');
  });

  it('Status Sould be Cold', () => {
    const { getByTestId } = renderWithRedux(<Container {...defaultProps} currentTemperature={-2} />);
    expect(getByTestId('status-label')).toHaveTextContent('Cold');
  });

  it('Status Sould be Normal', () => {
    const { getByTestId } = renderWithRedux(<Container {...defaultProps} />);
    const el=getByTestId('card');
    expect(el).toHaveClass('card');
    expect(el).not.toHaveClass('hot');
    expect(el).not.toHaveClass('cold');
  });

  it('Cold temperature has cold css class', () => {
    const { getByTestId } = renderWithRedux(<Container {...defaultProps} currentTemperature={-2} />);
    const el=getByTestId('card');
    expect(el).toHaveClass('card');
    expect(el).not.toHaveClass('hot');
    expect(el).toHaveClass('cold');
  });

  it('Cold temperature has Hcold css class', () => {
    const { getByTestId } = renderWithRedux(<Container {...defaultProps} currentTemperature={4} />);
    const el=getByTestId('card');
    expect(el).toHaveClass('card');
    expect(el).toHaveClass('hot');
    expect(el).not.toHaveClass('cold');
  });

});
