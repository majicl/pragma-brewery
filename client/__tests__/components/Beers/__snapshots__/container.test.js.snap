// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Container /> matches snapshot on responsive 1`] = `
<div
  class="card"
  data-testid="card"
>
  <div
    class="card-body text-center"
  >
    <h5
      class="mb-2 text-dark font-weight-normal"
    />
    <h2
      class="mb-4 text-dark font-weight-bold"
    >
      1
      °C
    </h2>
    <div
      class="dashboard-progress dashboard-progress-1 d-flex align-items-center justify-content-center item-parent"
    >
      <i
        class="mdi mdi-oil-temperature icon-md absolute-center text-dark"
      />
    </div>
    <p
      class="mt-4 mb-0 status-label"
      data-testid="status-label"
    >
      Normal
    </p>
    <h4
      class="mb-0 mt-2 text-dark"
    >
      <span>
        Condition:
      </span>
      -1
      °C
      <span>
         - 
      </span>
      3
      °C
    </h4>
  </div>
</div>
`;
